rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,528 @@
++}
+
+message MFAAuthenticateResponse {
+  oneof Response {
+    TOTPResponse TOTP = 2;
+    webauthn.CredentialAssertionResponse Webauthn = 3;
++   PrivilegeTokenResponse Token = 4;
+  }
+}
+```
+
+### Backward Compatibility
+
+It's possible that an old client or proxy version could result in an attempted
+SSO login using an auth connector setup for MFA only. The Auth server will
+check auth requests against the configured auth connector to ensure that these
+login attempts are prevented. Checks will be added to the following endpoints:
+
+- `rpc CreateSAMLAuthRequest`
+- `rpc CreateOIDCAuthRequest`
+- `rpc CreateGithubAuthRequest`
+- `http /v1/saml/requests/validate`
+- `http /v1/oidc/requests/validate`
+- `http /v1/github/requests/validate`
+
+### Audit Events
+
+SSO MFA requests will be tracked through the existing audit events that contain
+`MFADeviceMetadata`. Since SSO as MFA doesn't correspond to actual user MFA
+devices in the backend, this metadata will be picked from the auth connector:
+
+- `DeviceName` - Name of the auth connector
+- `DeviceID` - UUID of the auth connector
+- `DeviceType` - SSO type (`OIDC`, `SAML`, or `Github`)
+
+```proto
+message MFADeviceMetadata {
+  // Name is the user-specified name of the MFA device.
+  string DeviceName = 1 [(gogoproto.jsontag) = "mfa_device_name"];
+  // ID is the UUID of the MFA device generated by Teleport.
+  string DeviceID = 2 [(gogoproto.jsontag) = "mfa_device_uuid"];
+  // Type is the type of this MFA device.
+  string DeviceType = 3 [(gogoproto.jsontag) = "mfa_device_type"];
+}
+```
+
+#### Privilege Tokens
+
+When a privilege token is created, the context of its creation is lost. This
+makes it difficult to tie audit event between the SSO login that creates the
+token and the Per-session MFA certificate issuance that consumes the token.
+
+To amend this, we will store the details of the MFA device used to create the
+privilege token in the backend token resource. This data can then be included
+in the `privilege_token.create` and `mfa_auth_challenge.validate` events.
+
+### Additional considerations
+
+#### Temporary SSO MFA device in backend
+
+There are a few nice to have features that could be implemented if we created
+temporary SSO MFA devices for users. For example, users would be able to list
+SSO MFA devices available to them.
+
+```console
+> tsh mfa ls
+Name     Type     Added at Last used                     
+-------- -------- -------- ---------
+yubi     WebAuthn ...      ...
+okta-mfa SAML     ...      ...
+```
+
+We could then implement a way for clients to choose an SSO MFA device by
+providing new flags such as `--mfa-auth-type` and `--mfa-auth-name`.
+
+However, this idea would require significant engineering to handle
+automatic creation, automatic deletion, user interaction (tsh mfa rm),
+and other edge cases.
+
+For now, users will always be prompted to use the default MFA SSO connector,
+as configured in the cluster auth preference.
+
+#### OIDC ACR Values
```

Comment:
We could set these automatically for commonly-used providers. Okta is probably worth supporting, for example.

WDYT?

Reply:
I still think it would be best left to the cluster administrator. For example, with Okta, automatically setting `acr_values=phr` may be too restrictive for some use cases, and setting `acr_values=urn:okta:loa:2fa:any:ifpossible` may be too lenient in other cases. As long as we support a way to configure it, I think that's sufficient.

Reply:
Fair enough.